### PR TITLE
Fix kubernetes_service nil ptr dereference

### PR DESCRIPTION
**Pull request by lxea**
Based off of the stack trace in #9721 (havent successfuly reproduced the issue), it appears that the nil ptr dereference happens when `creds.tlsConfig` is nil. 

#9721

### CODE REVIEW

File: lib/kube/proxy/forwarder.go [hunk 1]
```diff
@@ -1502,6 +1502,9 @@ func (f *Forwarder) newClusterSessionLocal(ctx authContext) (*clusterSession, er
 	if !ok {
 		return nil, trace.NotFound("kubernetes cluster %q not found", ctx.kubeCluster)
 	}
+	if creds.tlsConfig == nil {
```
Comment (codingllama) [inline lib/kube/proxy/forwarder.go:1505]: At a glance this seems reasonable, but I'm more concerned about how we got a "bad" cred into the creds map in the first place. Maybe we should try and stop that from happening.

**Comment by russjones** [comment]
@codingllama @jimbishopp Can you guys take a look?

**Comment by codingllama** [comment]
> @codingllama @jimbishopp Can you guys take a look?

I have an open question above, but happy to approve if it's too much trouble.

**Comment by jimbishopp** [comment]
@russjones I was waiting for Alan's question to be answered as well. I'm was also concerned that it cannot be reproduced.

**Comment by jimbishopp** [comment]
@lxea LMK if this is your understanding of this issue.

@codingllama This may answer the question.

It looks like the k8s lib is bypassing the TLS config under certain conditions resulting in the TLS config being set to `nil`. 

`k8s.io/client-go/transport/transport.go:60`
```go
 func TLSConfigFor(c *Config) (*tls.Config, error) {
	if !(c.HasCA() || c.HasCertAuth() || c.HasCertCallback() || c.TLS.Insecure || len(c.TLS.ServerName) > 0 || len(c.TLS.NextProtos) > 0) {
		return nil, nil
	}
...
}
```

We then panic while writing a debug log message:
`kube/proxy/forwarder.go:1499`

```go    
func (f *Forwarder) newClusterSessionLocal(ctx authContext) (*clusterSession, error) {
...
     f.log.Debugf("local Servername: %v", creds.tlsConfig.ServerName)
```



I don't know enough about our k8s implementation yet, but I assume by the way this is written that client TLS is not required? If so, is failing to start the forwarder the right move here?

**Comment by lxea** [comment]
> @lxea LMK if this is your understanding of this issue.
> ....
Yeah, this is my current understanding however I wasnt able to reproduce the error unfortunatly.

**Comment by russjones** [comment]
@r0mant What do you think? Should we merge this without a repro?

**Comment by r0mant** [comment]
It may be fine to merge without the repro but I'm not sure the current fix checks for nil in the best place. Looking at the code, it seems like `tlsConfig` is returned as nil in the extractNewCreds function from `rest.TLSConfigFor`:

https://github.com/gravitational/teleport/blob/v7.3.2/lib/kube/proxy/auth.go#L159

I think that would be a more appropriate place to check that tlsConfig is not nil (in addition to checking err).

As for the repro, I think getting a kubeconfig from the user experiencing this issue will help. Looking at rest.TLSConfigFor implementation (it's a method from K8s client lib) it returns "nil, nil" in a few specific cases based on the kubeconfig as @jimbishopp mentioned above.

File: lib/kube/proxy/auth.go [hunk 2]
```diff
@@ -174,6 +174,9 @@ func extractKubeCreds(ctx context.Context, cluster string, clientCfg *rest.Confi
 	if err != nil {
 		return nil, trace.Wrap(err, "failed to generate TLS config from kubeconfig: %v", err)
 	}
+	if tlsConfig == nil {
+		return nil, trace.Errorf("tlsConfig was nil")
```
Comment (r0mant) [inline lib/kube/proxy/auth.go:178]: Nit: We don't usually use trace.Errorf directly. Also, can we add more diagnostic information to the error message? For example, look at which cases TLSConfigFor can return a nil tlsConfig and log appropriate fields so the logs can help narrow down the issue if it happens again?
```suggestion
		return nil, trace.BadParameter("failed to generate TLS config from kubeconfig xxxxx") 
```

**Review by r0mant — COMMENTED** [review]
Also please add test coverage.

**Review by r0mant — APPROVED** [review]

**Comment by r0mant** [comment]
@codingllama @jimbishopp Can one of you give this another look please? We couldn't get the customer to share their kubeconfig with us so far so we want to merge this so it fixes the panic and logs the config (with sensitive fields redacted) which would help in troubleshooting if it happens again.

File: lib/kube/proxy/auth.go [hunk 3]
```diff
@@ -174,6 +174,19 @@ func extractKubeCreds(ctx context.Context, cluster string, clientCfg *rest.Confi
 	if err != nil {
 		return nil, trace.Wrap(err, "failed to generate TLS config from kubeconfig: %v", err)
 	}
+	if tlsConfig == nil {
+		cc := rest.CopyConfig(clientCfg)
```
Comment (codingllama) [inline lib/kube/proxy/auth.go:178]: Do we really need to echo the entire config? It seems a bit dangerous to do.

Maybe use rest.AnonymousClientConfig? Although I'd rather not echo everything if possible.

File: lib/kube/proxy/auth_test.go [hunk 4]
```diff
@@ -269,3 +270,17 @@ current-context: foo
 		})
 	}
 }
+
+func TestExtractKubeCreds(t *testing.T) {
+	t.Parallel()
+	_, err := extractKubeCreds(context.TODO(),
```
[the remark this answers is not in the document]
Comment (lxea) [reply lib/kube/proxy/auth_test.go:276]: I moved this test to be part of the above TestGetKubeCreds test, however I realized that this test is now failing as kubeCreds is now always empty as tlsConfig is never configured in this test, so I'm not sure if returning nil is right in auth.go now or if instead a default tls config should be used.

If it shouldn't, where would I find docs on how to configure tls config for this test?

File: lib/kube/proxy/auth_test.go [hunk 5]
```diff
@@ -252,11 +258,26 @@ current-context: foo
...
+			kubeconfigPath:     kubeconfigPath,
+			impersonationCheck: alwaysSucceeds,
+			assertErr: func(tt require.TestingT, err error, i ...interface{}) {
+				findErr := "failed to generate TLS config from kubeConfig. clientConfig"
+				for sc.Scan() {
+					if strings.Contains(sc.Text(), findErr) {
+						return
+					}
+				}
+				require.Fail(tt, "Failed to find error", "Failed to find error %q in the logs", findErr)
```
Comment (codingllama) [inline lib/kube/proxy/auth_test.go:273]: Simplify? require.Fail is a bit pointless, IMO, plus the stuttering message is weird.

```suggestion
				tt.Fatalf("Failed to find error %q in the logs", findErr)
```

File: lib/kube/proxy/auth_test.go [hunk 6]
```diff
@@ -252,11 +258,26 @@ current-context: foo
 				},
 			},
 			assertErr: require.NoError,
+		}, {
+			desc:               "kubernetes_service, bad kube creds",
+			serviceType:        KubeService,
+			kubeconfigPath:     kubeconfigPath,
+			impersonationCheck: alwaysSucceeds,
```
Comment (codingllama) [inline lib/kube/proxy/auth_test.go:265]: The setup looks correct, why would this fail?

File: lib/kube/proxy/auth_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (codingllama) [reply lib/kube/proxy/auth_test.go]: Would testing this in a separate method, with a different setup, help?

Let's see what Roman says.

File: lib/kube/proxy/auth_test.go (hunk 6 above)
Comment (lxea) [reply lib/kube/proxy/auth_test.go:265]: This, aswell as the other cases that have a `wants:` section that isnt empty fails as the changes in auth.go returns early if there isnt a tlsConfig so kubeCreds is always nil in these tests.

File: lib/kube/proxy/auth.go [hunk 7]
```diff
@@ -174,6 +174,19 @@ func extractKubeCreds(ctx context.Context, cluster string, clientCfg *rest.Confi
...
+		cc := rest.AnonymousClientConfig(clientCfg)
+		if len(cc.CAData) != 0 {
+			cc.CAData = []byte("REDACTED")
+		}
+		if len(cc.CertData) != 0 {
+			cc.CertData = []byte("REDACTED")
+		}
+		if len(cc.KeyData) != 0 {
+			cc.KeyData = []byte("REDACTED")
+		}
```
Comment (r0mant) [inline lib/kube/proxy/auth.go:187]: @lxea Is this still needed with `AnonymousClientConfig` being used?

Comment (lxea) [reply lib/kube/proxy/auth.go:187]: Only CAData is still needed here with AnonymousClientConfig

File: lib/kube/proxy/auth_test.go [hunk 8]
```diff
@@ -125,15 +130,57 @@ func failsForCluster(clusterName string) ImpersonationPermissionsChecker {
 	}
 }
 
+func createFileWithContents(t *testing.T, contents []byte) string {
+	tmpFile, err := os.CreateTemp("", "teleport")
```
Comment (codingllama) [inline lib/kube/proxy/auth_test.go:134]: FYI, you can use `t.TempDir()` as the basis for the path. Cleanup is automatic.

File: lib/kube/proxy/auth_test.go [hunk 9]
```diff
@@ -125,15 +130,57 @@ func failsForCluster(clusterName string) ImpersonationPermissionsChecker {
...
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		os.Remove(tmpFile.Name())
+	})
+	_, err = tmpFile.Write(contents)
```
Comment (codingllama) [inline lib/kube/proxy/auth_test.go:139]: We need an explicit close after the Write, plus we also need to make sure the file closes in other scenarios. 

I suggest you use `os.WriteFile`, it sidesteps all of those.

File: lib/kube/proxy/auth_test.go [hunk 10]
```diff
@@ -125,15 +130,57 @@ func failsForCluster(clusterName string) ImpersonationPermissionsChecker {
...
+
 func TestGetKubeCreds(t *testing.T) {
 	ctx := context.TODO()
 	const teleClusterName = "teleport-cluster"
 
-	tmpFile, err := os.CreateTemp("", "teleport")
+	cert, err := utils.GenerateSelfSignedCert([]string{"localhost"})
+	require.NoError(t, err)
+	certFile := createFileWithContents(t, cert.Cert)
+	keyFile := createFileWithContents(t, cert.PrivateKey)
```
Comment (codingllama) [inline lib/kube/proxy/auth_test.go:151]: Building on the comment above, write all test files to the same temp dir?

File: lib/kube/proxy/auth_test.go [hunk 11]
```diff
@@ -157,8 +204,11 @@ users:
 - name: creds
 current-context: foo
 `))
-	require.NoError(t, err)
-	require.NoError(t, tmpFile.Close())
+
+	logger := utils.NewLoggerForTests()
+	buf := bytes.NewBuffer([]byte{})
+	logger.SetOutput(buf)
```
Comment (codingllama) [inline lib/kube/proxy/auth_test.go:210]:
```suggestion
	logger.SetOutput(io.Discard)
```

**Review by codingllama — APPROVED** [review]

**Pull request merged**